File: ios/Classes/CourierClientMethodHandler.swift
//
//  CourierClientMethodHandler.swift
//  courier_flutter
//
//  Created by Michael Miller on 8/2/24.
//

import Foundation
import Courier_iOS

internal class CourierClientMethodHandler: CourierFlutterMethodHandler, FlutterPlugin {
    
    private var clients: [String: CourierClient] = [:]
    
    static func getChannel(with registrar: FlutterPluginRegistrar) -> FlutterMethodChannel {
        return FlutterMethodChannel(name: CourierFlutterChannel.client.rawValue, binaryMessenger: registrar.messenger())
    }
    
    static func register(with registrar: any FlutterPluginRegistrar) {
        registrar.addMethodCallDelegate(
            CourierClientMethodHandler(),
            channel: CourierClientMethodHandler.getChannel(with: registrar)
        )
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        Task {
            
            do {
                
                guard let params = call.arguments as? Dictionary<String, Any> else {
                    throw CourierFlutterError.invalidParameter(value: "params")
                }
                
                switch call.method {
                    
                    // MARK: Client
                    
                case "client.add":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    do {
                        let existingClient = try getClient(for: clientId)
                        result(clientId)
                    } catch {
                        let newClient = try params.toClient()
                        clients[clientId] = newClient
                        result(clientId)
                    }
                    
                case "client.remove":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    clients.removeValue(forKey: clientId)
                    
                    result(clientId)
                    
                    // MARK: Brand
                    
                case "brands.get_brand":
                    
                    let (clientId, brandId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("brandId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let brand = try await client.brands.getBrand(
                        brandId: brandId
                    )
                    
                    let json = try brand.toJson()
                    result(json)
                    
                    // MARK: Token Management
                    
                case "tokens.put_user_token":
                    
                    let (clientId, token, provider): (String, String, String) = (
                        try params.extract("clientId"),
                        try params.extract("token"),
                        try params.extract("provider")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let deviceParams = params["device"] as? [String: Any]
                    let device = try deviceParams?.toCourierDevice()
                    
                    try await client.tokens.putUserToken(
                        token: token,
                        provider: provider,
                        device: device ?? CourierDevice()
                    )
                    
                    result(nil)
                    
                case "tokens.delete_user_token":
                    
                    let (clientId, token): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("token")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    try await client.tokens.deleteUserToken(
                        token: token
                    )
                    
                    result(nil)
                    
                    // MARK: Preferences
                    
                case "preferences.get_user_preferences":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let paginationCursor = params["paginationCursor"] as? String
                    
                    let res = try await client.preferences.getUserPreferences(
                        paginationCursor: paginationCursor
                    )
                    
                    let json = try res.toJson()
                    result(json)
                    
                case "preferences.get_user_preference_topic":
                    
                    let (clientId, topicId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("topicId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let res = try await client.preferences.getUserPreferenceTopic(
                        topicId: topicId
                    )
                    
                    let json = try res.toJson()
                    result(json)
                    
                case "preferences.put_user_preference_topic":
                    
                    let (clientId, topicId, status, hasCustomRouting, customRouting): (String, String, String, Bool, [String]) = (
                        try params.extract("clientId"),
                        try params.extract("topicId"),
                        try params.extract("status"),
                        try params.extract("hasCustomRouting"),
                        try params.extract("customRouting")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    try await client.preferences.putUserPreferenceTopic(
                        topicId: topicId,
                        status: CourierUserPreferencesStatus(rawValue: status) ?? .unknown,
                        hasCustomRouting: hasCustomRouting,
                        customRouting: customRouting.map { CourierUserPreferencesChannel(rawValue: $0) ?? .unknown }
                    )
                    
                    result(nil)
                    
                    // MARK: Inbox
                    
                case "inbox.get_messages":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let paginationLimit = params["paginationLimit"] as? Int
                    let startCursor = params["startCursor"] as? String

                    let res = try await client.inbox.getMessages(
                        paginationLimit: paginationLimit ?? Courier.shared.inboxPaginationLimit,
                        startCursor: startCursor
                    )

                    let json = try res.toJson()
                    result(json)

                case "inbox.get_archived_messages":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let paginationLimit = params["paginationLimit"] as? Int
                    let startCursor = params["startCursor"] as? String

                    let res = try await client.inbox.getArchivedMessages(
                        paginationLimit: paginationLimit ?? Courier.shared.inboxPaginationLimit,
                        startCursor: startCursor
                    )

                    let json = try res.toJson()
                    result(json)

                case "inbox.get_unread_message_count":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    let count = try await client.inbox.getUnreadMessageCount()
                    result(count)

                case "inbox.get_message_by_id":
                    
                    let (clientId, messageId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId")
                    )
                    
                    let client = try getClient(for: clientId)

                    let res = try await client.inbox.getMessage(
                        messageId: messageId
                    )

                    let json = try res.toJson()
                    result(json)

                case "inbox.click_message":
                    
                    let (clientId, messageId, trackingId): (String, String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId"),
                        try params.extract("trackingId")
                    )
                    
                    let client = try getClient(for: clientId)

                    try await client.inbox.click(
                        messageId: messageId,
                        trackingId: trackingId
                    )

                    result(nil)

                case "inbox.unread_message":
                    
                    let (clientId, messageId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId")
                    )
                    
                    let client = try getClient(for: clientId)

                    try await client.inbox.unread(
                        messageId: messageId
                    )

                    result(nil)

                case "inbox.read_message":
                    
                    let (clientId, messageId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId")
                    )
                    
                    let client = try getClient(for: clientId)

                    try await client.inbox.read(
                        messageId: messageId
                    )

                    result(nil)

                case "inbox.open_message":
                    
                    let (clientId, messageId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId")
                    )
                    
                    let client = try getClient(for: clientId)

                    try await client.inbox.open(
                        messageId: messageId
                    )

                    result(nil)

                case "inbox.archive_message":
                    
                    let (clientId, messageId): (String, String) = (
                        try params.extract("clientId"),
                        try params.extract("messageId")
                    )
                    
                    let client = try getClient(for: clientId)

                    try await client.inbox.archive(
                        messageId: messageId
                    )

                    result(nil)

                case "inbox.read_all_messages":
                    
                    let (clientId): (String) = (
                        try params.extract("clientId")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    try await client.inbox.readAll()
                    result(nil)
                    
                    // MARK: Tracking
                    
                case "tracking.post_tracking_url":
                    
                    let (clientId, url, event): (String, String, String) = (
                        try params.extract("clientId"),
                        try params.extract("url"),
                        try params.extract("event")
                    )
                    
                    let client = try getClient(for: clientId)
                    
                    guard let trackingEvent = CourierTrackingEvent(rawValue: event) else {
                        throw CourierFlutterError.missingParameter(value: "tracking_event")
                    }
                    
                    try await client.tracking.postTrackingUrl(
                        url: url,
                        event: trackingEvent
                    )
                    
                    result(nil)
                    
                default:
                    
                    result(FlutterMethodNotImplemented)
                    
                }
                
            } catch {
                
                result(error.toFlutterError())
                
            }
            
        }
          
    }
    
    private func getClient(for clientId: String) throws -> CourierClient {
        
        guard let client = clients[clientId] else {
            throw CourierFlutterError.invalidParameter(value: "clientId")
        }
        
        return client
        
    }
    
}

File: android/src/main/kotlin/com/courier/courier_flutter/CourierClientMethodHandler.kt
package com.courier.courier_flutter

import com.courier.android.Courier
import com.courier.android.models.CourierDevice
import com.courier.android.models.CourierPreferenceChannel
import com.courier.android.models.CourierPreferenceStatus
import com.courier.android.models.CourierTrackingEvent
import com.courier.android.modules.inboxPaginationLimit
import com.courier.courier_flutter.CourierPlugin.Companion.TAG
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler

internal class CourierClientMethodHandler : MethodCallHandler {

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) = post {

        try {

            val params = call.arguments as? HashMap<*, *>

            val client = params?.toClient() ?: throw MissingParameter("client")

            when (call.method) {

                // == Brand ==

                "client.brands.get_brand" -> {

                    val (brandId) = listOf<String>(
                        params.extract("brandId"),
                    )

                    val brand = client.brands.getBrand(brandId)
                    val json = brand.toJson()
                    result.success(json)

                }

                // == Token Management ==

                "client.tokens.put_user_token" -> {

                    val (token, provider) = listOf<String>(
                        params.extract("token"),
                        params.extract("provider"),
                    )

                    val deviceParams = params["device"] as? Map<*, *>
                    val device = deviceParams?.toCourierDevice()

                    client.tokens.putUserToken(
                        token = token,
                        provider = provider,
                        device = device ?: CourierDevice.current,
                    )

                    result.success(null)

                }

                "client.tokens.delete_user_token" -> {

                    val (token) = listOf<String>(
                        params.extract("token"),
                    )

                    client.tokens.deleteUserToken(
                        token = token,
                    )

                    result.success(null)

                }

                // == Inbox ==

                "client.inbox.get_messages" -> {

                    val paginationLimit = params["paginationLimit"] as? Int
                    val startCursor = params["startCursor"] as? String

                    val res = client.inbox.getMessages(
                        paginationLimit = paginationLimit ?: Courier.shared.inboxPaginationLimit,
                        startCursor = startCursor
                    )

                    val json = res.toJson()
                    result.success(json)

                }

                "client.inbox.get_archived_messages" -> {

                    val paginationLimit = params["paginationLimit"] as? Int
                    val startCursor = params["startCursor"] as? String

                    val res = client.inbox.getArchivedMessages(
                        paginationLimit = paginationLimit ?: Courier.shared.inboxPaginationLimit,
                        startCursor = startCursor
                    )

                    val json = res.toJson()
                    result.success(json)

                }

                "client.inbox.get_unread_message_count" -> {

                    val count = client.inbox.getUnreadMessageCount()
                    result.success(count)

                }

                "client.inbox.get_message_by_id" -> {

                    val (messageId) = listOf<String>(
                        params.extract("messageId"),
                    )

                    val res = client.inbox.getMessage(
                        messageId = messageId
                    )

                    val json = res.toJson()
                    result.success(json)

                }

                "client.inbox.click_message" -> {

                    val (messageId, trackingId) = listOf<String>(
                        params.extract("messageId"),
                        params.extract("trackingId"),
                    )

                    client.inbox.trackClick(
                        messageId = messageId,
                        trackingId = trackingId
                    )

                    result.success(null)

                }

                "client.inbox.unread_message" -> {

                    val (messageId) = listOf<String>(
                        params.extract("messageId"),
                    )

                    client.inbox.trackUnread(
                        messageId = messageId,
                    )

                    result.success(null)

                }

                "client.inbox.read_message" -> {

                    val (messageId) = listOf<String>(
                        params.extract("messageId"),
                    )

                    client.inbox.trackRead(
                        messageId = messageId,
                    )

                    result.success(null)

                }

                "client.inbox.open_message" -> {

                    val (messageId) = listOf<String>(
                        params.extract("messageId"),
                    )

                    client.inbox.trackOpened(
                        messageId = messageId,
                    )

                    result.success(null)

                }

                "client.inbox.archive_message" -> {

                    val (messageId) = listOf<String>(
                        params.extract("messageId"),
                    )

                    client.inbox.trackArchive(
                        messageId = messageId,
                    )

                    result.success(null)

                }

                "client.inbox.read_all_messages" -> {

                    client.inbox.trackAllRead()

                    result.success(null)

                }

                // == Preferences ==

                "client.preferences.get_user_preferences" -> {

                    val paginationCursor = params["paginationCursor"] as? String

                    val res = client.preferences.getUserPreferences(
                        paginationCursor = paginationCursor
                    )

                    val json = res.toJson()
                    result.success(json)

                }

                "client.preferences.get_user_preference_topic" -> {

                    val (topicId) = listOf<String>(
                        params.extract("topicId"),
                    )

                    val res = client.preferences.getUserPreferenceTopic(
                        topicId = topicId
                    )

                    val json = res.toJson()
                    result.success(json)

                }

                "client.preferences.put_user_preference_topic" -> {

                    val topicId = params.extract<String>("topicId")
                    val status = params.extract<String>("status")
                    val hasCustomRouting = params.extract<Boolean>("hasCustomRouting")
                    val customRouting = params.extract<List<String>>("customRouting")

                    client.preferences.putUserPreferenceTopic(
                        topicId = topicId,
                        status = CourierPreferenceStatus.valueOf(status),
                        hasCustomRouting = hasCustomRouting,
                        customRouting = customRouting.map { CourierPreferenceChannel.fromString(it) },
                    )

                    result.success(null)

                }

                // == Tracking ==

                "client.tracking.post_tracking_url" -> {

                    val (url, event) = listOf<String>(
                        params.extract("url"),
                        params.extract("event"),
                    )

                    client.tracking.postTrackingUrl(
                        url = url,
                        event = CourierTrackingEvent.valueOf(event)
                    )

                    result.success(null)

                }

                else -> {
                    result.notImplemented()
                }

            }

        } catch (e: Exception) {

            result.error(TAG, e.message, e)

        }

    }

}

internal fun Map<*, *>.toCourierDevice(): CourierDevice {
    val appId = this["app_id"] as? String
    val adId = this["ad_id"] as? String
    val deviceId = this["device_id"] as? String
    val platform = this["platform"] as? String
    val manufacturer = this["manufacturer"] as? String
    val model = this["model"] as? String
    return CourierDevice(
        app_id = appId,
        ad_id = adId,
        device_id = deviceId,
        platform = platform,
        manufacturer = manufacturer,
        model = model
    )
}